### DISCUSSION TITLE
Kafka Redpanda container not stopped and delete on application end

### DISCUSSION THREAD

**Discussion by ppatierno** [quarkusio/quarkus#22532] [Q&A] [answered]
While developing an application using Kafka Streams using then Kafka Dev Services, we know that a Redpanda based Kafka broker is started in a container:

```shell
CONTAINER ID   IMAGE                         COMMAND                  CREATED        STATUS                      PORTS                                                   NAMES
634a0bfbe491   vectorized/redpanda:v21.5.5   "sh -c 'while [ ! -f…"   7 hours ago    Up 7 hours                  9644/tcp, 0.0.0.0:49154->9092/tcp, :::49154->9092/tcp   hungry_heisenberg
```

When is it supposed to be stopped and removed? When I stop my application it doesn't happen.
Is it related to the fact that I had to disable ryuk because of this other problem described here https://github.com/quarkusio/quarkus/discussions/22524 ?

[a comment by an automated account is not shown]

**Reply by cescoffier** [reply]
Yes, if you disable Riuk then the watchdog does not catch the remaining containers.

**Reply by ppatierno** [reply]
@cescoffier so it means that what is suggested in #22524 is not a good one in the end and that the overall developer UX is not that great.
If ryuk doesn't work for you, you can disable it but then Redpanda container is left running ... so your only solution is about having your own Kafka cluster running avoiding to run Redpanda but then it's not what Quarkus wants to achieve for dev cycle.
Any other better idea about not using ryuk or having it working, or anything better than how it works currently?

**Reply by cescoffier** [reply]
Not really, this is how dev services using test containers are intended to work. 

I'm also curious about your setup as I tried on 3 different OS and I didn't run into any issue (just added the kafka client extension).   I use docker (let me know if you use podman), and for my Linux it's a fresh Ubuntu.

**Reply by ppatierno** [reply]
I have docker on Fedora.

**Reply by ppatierno** [reply]
But as I mentioned on the other issue, who really likes to have a container running as a privileged container with rights of the current host? Tbh, I have never used a privileged container but AFAIU this is how ryuk runs.

**Reply by ppatierno** [reply]
I tried again this morning and even disabling ryuk (which doesn't work on my side, #22524) now the redpanda container is deleted when I stop the application. Tbh I don't know why, but it's cleaned even without ryuk. Is that expected somehow?

**Reply by cescoffier** [reply]
So, when we have a clean termination of the dev mode, we stop the container. However, there are cases where the clean-up cannot be executed (like a hard crash). In this case, ruyk is doing the cleanup.

**Reply by ppatierno** [reply]
Ok it makes more sense. 

> In this case, ruyk is doing the cleanup.

I am still not convinced that a dev wants a privileged container running to do the clean up, or at least not me. Anyway I still have the need to disable it otherwise my app doesn't work with dev services :-(

**Comment by ozangunalp** [comment, marked as answer]
Let me clarify two things here: 

1/ Test containers does not run ryuk container in privileged mode by default. That is the reason it didn't work for you in the first place. Seems like there is a particular issue with Fedora: https://github.com/testcontainers/testcontainers-java/issues/4014. I personally would qualify this as a non-issue as it doesn't impact most people and if it does there are multiple workarounds.

2/ There is actually two safeguards for stopping and deleting the dev service container. The first one is on the Quarkus end: we close and delete the container on the application exit. Second one is on the test containers end, with or without ryuk, it registers a runtime hook for cleaning up the container. Of course, if the process quits abruptly, (ex. with `kill -9`) the only thing that would work is ryuk which acts as a watchdog.

**Reply by ppatierno** [reply]
Thanks for the detailed explanation!
